winpr/libwinpr/sspi/Kerberos/kerberos.c

```diff
@@ -659,117 +659,154 @@ static BOOL kerberos_mk_tgt_token(SecBuffer* buf, int msg_type, char* sname, cha
+	char* ndst = strncpy(dst, src, dstSize);
+	if (ndst < dst)
+		return SIZE_MAX;
+	return ndst - dst;
+}
 
-	/* KERB-TGT-REQUEST (SEQUENCE) */
-	if (!WinPrAsn1DecReadSequence(&dec, &dec2))
+static BOOL kerberos_rd_tgt_req(WinPrAsn1Decoder* dec, char** target)
+{
+	BOOL rc = FALSE;
+	char* str = NULL;
+	if (!target)
 		return FALSE;
-	dec = dec2;
+	*target = NULL;
 
-	/* pvno [0] INTEGER */
-	if (!WinPrAsn1DecReadContextualInteger(&dec, 0, &error, &val) || val != 5)
-		return FALSE;
+	wStream s = WinPrAsn1DecGetStream(dec);
+	const size_t len = Stream_Length(&s);
+	if (len == 0)
+		return TRUE;
 
-	/* msg-type [1] INTEGER */
-	if (!WinPrAsn1DecReadContextualInteger(&dec, 1, &error, &val))
+	size_t offset = 0;
+	char* buf = calloc(len + 1, sizeof(char));
+	if (!buf)
 		return FALSE;
 
-	if (val == KRB_TGT_REQ)
-	{
-		if (!target)
-			return FALSE;
-		*target = NULL;
-
-		s = WinPrAsn1DecGetStream(&dec);
-		len = Stream_Length(&s);
-		if (len == 0)
-			return TRUE;
+	WinPrAsn1Decoder dec2 = { 0 };
+	WinPrAsn1_tagId tag = 0;
+	if (!WinPrAsn1DecReadContextualTag(dec, &tag, &dec2))
+		goto end;
 
-		buf = malloc(len);
-		if (!buf)
-			return FALSE;
+	if (tag == 2)
+	{
+		WinPrAsn1Decoder seq = { 0 };
+		/* server-name [2] PrincipalName (SEQUENCE) */
+		if (!WinPrAsn1DecReadSequence(&dec2, &seq))
+			goto end;
 
-		*buf = 0;
-		*target = buf;
+		/* name-type [0] INTEGER */
+		BOOL error = FALSE;
+		WinPrAsn1_INTEGER val = 0;
+		if (!WinPrAsn1DecReadContextualInteger(&seq, 0, &error, &val))
+			goto end;
 
-		if (!WinPrAsn1DecReadContextualTag(&dec, &tag, &dec2))
-			goto fail;
+		/* name-string [1] SEQUENCE OF GeneralString */
+		if (!WinPrAsn1DecReadContextualSequence(&seq, 1, &error, &dec2))
+			goto end;
 
-		if (tag == 2)
+		while (WinPrAsn1DecPeekTag(&dec2, &tag))
 		{
-			WinPrAsn1Decoder seq;
-			/* server-name [2] PrincipalName (SEQUENCE) */
-			if (!WinPrAsn1DecReadSequence(&dec2, &seq))
-				goto fail;
-
-			/* name-type [0] INTEGER */
-			if (!WinPrAsn1DecReadContextualInteger(&seq, 0, &error, &val))
-				goto fail;
-
-			/* name-string [1] SEQUENCE OF GeneralString */
-			if (!WinPrAsn1DecReadContextualSequence(&seq, 1, &error, &dec2))
-				goto fail;
+			char* lstr = NULL;
+			if (!WinPrAsn1DecReadGeneralString(&dec2, &lstr))
+				goto end;
 
-			while (WinPrAsn1DecPeekTag(&dec2, &tag))
+			if (buf != &buf[offset])
 			{
-				if (!WinPrAsn1DecReadGeneralString(&dec2, &str))
-					goto fail;
-
-				if (buf != *target)
-					*buf++ = '/';
-				buf = stpcpy(buf, str);
-				free(str);
+				offset += append(&buf[offset], "/", len - offset);
+				if (offset > len)
+					goto end;
 			}
 
-			if (!WinPrAsn1DecReadContextualTag(&dec, &tag, &dec2))
-				return TRUE;
+			offset += append(&buf[offset], lstr, len - offset);
+			if (offset > len)
+				goto end;
+			free(lstr);
 		}
 
-		/* realm [3] Realm */
-		if (tag != 3 || !WinPrAsn1DecReadGeneralString(&dec2, &str))
-			goto fail;
-
-		*buf++ = '@';
-		strcpy(buf, str);
```

Comment:
Is it me or the goal of all these changes is just to not have a `strcpy` which seems correctly checked and used in this function ? At least the new code is less trivial to read...

Reply:
Well on this one, the job was correctly done in terms of checks. And I had to look at the previous version to understand what the new code was doing.

Reply:
@hardening ok, split out the changes in `kerberos.c` to a separate pr as that is indeed quite a change.